### ISSUE TITLE
The most recent commit (49e2723) fails to build on Ubuntu.

### ISSUE THREAD

**Issue by mainframed** [SDL-Hercules-390/hyperion#396]
___
_**NOTE:** &nbsp;issues [\#392](https://github.com/SDL-Hercules-390/hyperion/issues/392)_ "WHY static inline functions in header files?" _and [\#399](https://github.com/SDL-Hercules-390/hyperion/issues/396)_ "warning: ‘fetch_dw_noswap’ is static but used in inline function ‘z900_vfetch8’ which is not static"&nbsp;_are both closely related to this issue._
___
The most recent commit (49e2723ea19d556473be372f3a14354f848bd18a) fails to build on Ubuntu. 

Linux Distro info:

```
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=20.04
DISTRIB_CODENAME=focal
DISTRIB_DESCRIPTION="Ubuntu 20.04.2 LTS"
$ gcc --version
gcc (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Copyright (C) 2019 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
$ make --version
GNU Make 4.2.1
Built for x86_64-pc-linux-gnu
Copyright (C) 1988-2016 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
```
Fresh build of ubuntu with the following installed `sudo apt-get install m4 make autoconf automake cmake flex build-essential regina-rexx libbz2-dev libregina3-dev zlib1g-dev unzip c3270 ncat`

Installed and built extra packages.

ran configure: `./configure --enable-cckd-bzip2 --enable-het-bzip2 --enable-regina-rexx --enable-extpkgs=/home/mainframed/sysgen/build/hercpkgs --enable-optimization="-O3 -march=native"`

Running `make` fails:

* [makelog.txt](https://github.com/SDL-Hercules-390/hyperion/files/6711357/makelog.txt):

Summary, multiple warnings appear that look like:

```
inline.h:32:20: warning: inline function 'add_logical' declared but never defined
   32 | extern inline int  add_logical       ( U32 *result, U32 op1, U32 op2 );
```

And it ultimately fails with:

```
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `add_logical'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `add_signed_long'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `add_logical_long'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `z900_fetch_doubleword_absolute'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `s390_fetch_main_absolute'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `s370_move_chars'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `s370_vfetch4'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `s390_vstorec'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `s370_vfetch8'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `z900_vstore2'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `s390_is_fetch_protected'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `s390_vfetchb'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `z900_fetch_fullword_absolute'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `s390_move_chars'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `z900_move_chars'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `sub_logical'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `mul_signed_long'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `s370_fetch_fullword_absolute'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `z900_vfetch2'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `z900_vfetchc'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `s370_vfetchc'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `s370_validate_operand'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `s390_vstore2'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `s390_vfetch2'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `s390_vstoreb'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `s390_fetch_fullword_absolute'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `s370_vfetchb'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `s370_fetch_halfword_absolute'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `z900_move_charx'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `s390_apply_prefixing'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `div_signed'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `s370_vstore8'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `s370_vstore4'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `s390_is_low_address_protected'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `z900_vfetchb'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `s370_is_store_protected'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `s370_vfetch2'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `z900_is_low_address_protected'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `s390_instfetch'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `z900_vstore4'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `mult_logical_long'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `z900_apply_prefixing'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `s390_vfetch8'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `sub_logical_long'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `mul_signed'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `s390_store_fullword_absolute'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `z900_vstoreb'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `s370_is_fetch_protected'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `s370_instfetch'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `s370_vstorec'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `z900_vfetch4'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `z900_vstore8'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `z900_is_store_protected'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `z900_store_doubleword_absolute'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `s390_is_store_protected'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `z900_move_chars_rl'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `s390_vstore4'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `z900_vfetch8'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `s390_vfetch4'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `s370_vstore2'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `add_signed'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `s390_vfetchc'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `s370_fetch_main_absolute'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `z900_validate_operand'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `s390_validate_operand'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `z900_store_fullword_absolute'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `z900_vstorec'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `sub_signed'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `s390_vstore8'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `z900_is_fetch_protected'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `s370_vstoreb'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `div_logical_long'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `concpy'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `s370_is_low_address_protected'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `sub_signed_long'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `z900_fetch_main_absolute'
/usr/bin/ld: ./.libs/libherc.so: undefined reference to `z900_instfetch'
```

_**Temporary workaround:**_

* Cloned hyperion to a seperate folder, then ran  `git checkout a6fb047bc1d13bfdffd57d45fb36d811a632412b` and after running the same configure/make, it built succesfully.

**Comment by Fish-Git** [comment]
This should now be fixed by commit 328266663492b8ce7d7473000f6ca441750421d4.

I tested it on my own KDE Neon User Edition 5.22 VMware virtual machine with kernel 5.8.0-59-generic, gcc (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0 and clang version 10.0.0-4ubuntu1, and it now builds clean.

@mainframed: &nbsp;Please refresh your repo (or re-clone) and try again to confirm.

Thanks!

**Comment by wrljet** [comment]
Fish,

I'm curious, what is the reason for renaming `logical_to_main` to `logical_to_main_l`?

Bill

**Comment by Fish-Git** [comment]
> ... what is the reason for renaming `logical_to_main` to `logical_to_main_l`?

Because the only thing the `logical_to_main` function was originally doing was simply calling `logical_to_main_l` anyway:

https://github.com/SDL-Hercules-390/hyperion/blob/2e9504e2416a2b82cf03b92a3605069a77c79fbb/dat.c#L2422-L2427

So why not just have everyone call `logical_to_main_l` _directly_ instead of having to jump through a do-nothing intermediate function?

**Comment by wrljet** [comment]
Fish,

While merging this into my in-progress work, I ran into some gcc warnings related to this inlining.
I was using an old gcc, so I checked the Travis logs to see what it was using and found the same warnings there.

All similar to:
```
In file included from ../opcode.h:699,
                 from ../hstructs.h:19,
                 from ../hercules.h:102,
                 from ../assist.c:26:
../machdep.h:769:28: warning: ‘store_dw_noswap’ is static but used in inline function ‘z900_vstore8’ which is not static
   #define store_dw(_p, _v) store_dw_noswap((_p), CSWAP64((_v)))
                            ^~~~~~~~~~~~~~~
```

FWIW, Google is awash in this warning.

**Comment by Fish-Git** [comment]
> `machdep.h:769:28: warning: ‘store_dw_noswap’ is static but used in inline function ‘z900_vstore8’ which is not static`

I'm aware of it, and have marked it for _eventual_ follow up. Fixing it will probably take a bit of time and effort which I don't wish to expend at the moment. I'm trying like hell to get my Storage Key changes/fix ready for release (which I've been working on for quite some time) so I don't want to introduce yet another distraction at this time. Once my skey fix has been released I'll come back to re-visit this issue (but probably after I touch base with the people I _was_ working with on our TXF issue before these distractions yanked me away).

Basically what I guess I'm saying Bill is, all in good time, all in good time. &nbsp; `:)`

**Comment by wrljet** [comment]
Cool.  I just wanted make sure we were aware of it.

**Comment by Fish-Git** [comment]
> machdep.h:769:28: warning: ‘store_dw_noswap’ is static but used in inline function ‘z900_vstore8’ which is not static

_**FYI**_ &nbsp;#<span>1</span>: &nbsp; GitHub Issue [#399 **"warning: ‘fetch_dw_noswap’ is static but used in inline function ‘z900_vfetch8’ which is not static"**](https://github.com/SDL-Hercules-390/hyperion/issues/399) created.

_**FYI**_ &nbsp;#<span>2</span>: &nbsp; _Apparently_ this error is the [likely cause of of my "New Storage Key handling" crashing and burning on Linux.](https://github.com/SDL-Hercules-390/hyperion/issues/400#issuecomment-872072944)

**Comment by Fish-Git** [comment]
I'm going to go ahead and close this issue at this point now (even though I haven't heard back from @mainframed yet) since I'm pretty sure his originally reported problem has indeed now been resolved.

Of course there are still _other_ problems still remaining to be resolved, all related to the original can of worms I opened with issue #392, but I'll get all those resolving all of them too in due time. _THIS_ specific issue however (inability to build) has now been successfully resolved IMO.

**Comment by Fish-Git** [comment]
_**FYI:**_ &nbsp;Commits 173bb145832e1e7cfb87cdfb354c9d8c70439cda and 38d6835b7460737a83aca078c259c11167a6ef63 (in addition to commit 328266663492b8ce7d7473000f6ca441750421d4 [mentioned in my earlier comment](https://github.com/SDL-Hercules-390/hyperion/issues/396#issuecomment-869224649)) should also be considered to be a fix for this issue too, as it addresses the problem identified in issue #399 (as mentioned by Bill in his [earlier comment above](https://github.com/SDL-Hercules-390/hyperion/issues/396#issuecomment-869888648)), which occurs whenever commit 328266663492b8ce7d7473000f6ca441750421d4 is applied.
